### ISSUE TITLE
DISCUSS CI pathway: Pull requests and review

### ISSUE THREAD

**Issue by fnriv** [cse110-sp23-group3/cse110-sp23-group3#32]
Malcolm has advised for more pull requests and a more solidified way of reviewing code. We need to discuss asap (perhaps on slack since it's more urgent?

general idea

1.  code (as single, mob, pair)
2.  make a request for assistance for testing
3.  update as need and recurse 2/3 as needed
4. once ready create pull request? for style review and last min checks
5. merge

i'm not the most familiar with this so any help is appreciated, but let's try to get this solidified by tuesday night/wednesday noon.

**Comment by nsyousef** [comment]
I think Faith's idea would be a good start. I guess we won't really know if it works for us until we really try it for ourselves. I feel like it would help if we had some rules set for branching though. Right now, it seems like people are making branches whenever they want to work on something because they don't fully understand what other people are doing on their branches and don't want to mess anything up.

My suggestion would be to have a branch for anything that is completed, either main or another branch if you want. The GitHub actions should run whenever a pull request is made to that branch. We could start by making pull requests to this branch for everything we have so far. Then we can do code reviews on what we have so far and push them to that branch.

When creating new features, we could have people create branches off of this final product branch where they work on their issues. We can have people mob/pair program/etc. on such branches as needed, as well as to help with writing tests. When committing to the branch, be sure to use the proper format so that the release notes generator can create release notes from the commits. Once that feature is done (i.e. it meets the specifications in the relevant issue(s), has sufficient unit tests, and has JSDocs in the code) make a pull request to the branch containing the final product (either main or another branch if we choose to use another branch). Assign people to code review. The GitHub action will run on the pull request and lint the code, generate the docs, and run the tests. The reviewer and the original developer will be able to see if the linting/tests failed and will be able to close the pull request, fix the issues, and make a new pull request. Once the pull request is approved, it can be merged with the final product branch. Relevant issues should be tagged in the pull request so they close automatically when the pull request is merged. An issue would not be considered closable until its code has been merged to main (or the final product branch) and all the tests pass/docs are generated, etc. Whenever we want to increment the version number and generate release notes, we can tag the latest commit in the final product branch with the current version number (e.g. `v1.2.3`) to generate release notes for that version.

Let me know what you think about this idea. I tried to make it follow along with what Faith had said above, but with more details on specifics of branching, etc.

**Comment by nsyousef** [comment]
I also saw Anh had left a suggestion in another thread: https://github.com/cse110-sp23-group3/cse110-sp23-group3/issues/26#issuecomment-1566561482

**Comment by phuanh004** [comment]
Also, during the conversation today between Nick and me about the coding format, I notice that our coding styles are all over the place and some do not follow the standardjs guideline. (e.g. semicolon). So, I do not know if you want to implement the `Husky`, pre-commit. 

- It could be annoying because we have to fix all our linter before committing. Otherwise, you cannot commit.
- Useful Note: https://www.gitkraken.com/blog/husky-git-hooks

**Comment by henryzhang03** [comment]
We are currently using the following eslint and prettier configs, and I recently changed the Github Actions workflow to automatically correct mistakes, so the code style should be consistent:

.prettierrc:
```
{
  "semi": true,
  "singleQuote": true,
  "endOfLine":"auto"
}
```

.eslintrc.json:
```
{
  "extends": ["standard", "prettier"],
  "plugins": ["prettier"],
  "rules": {
    "prettier/prettier": [
      "error",
      {
        "endOfLine": "auto"
      }
    ]
  },
  "env": {
    "node": true
  }
}
```

**Comment by phuanh004** [comment]
@henryzhang03 Looks good! 
Since we go with standardjs, semi should be `never` tho!

**Comment by henryzhang03** [comment]
@phuanh004 I understand, but I kept that specifically on as the intention because of what Professor Powell stated about semicolons in JS. I personally agree with him as well, so we are using basically https://github.com/standard/semistandard.

**Comment by phuanh004** [comment]
@henryzhang03 Got cha!

**Comment by henryzhang03** [comment]
resolved.
